### PR TITLE
implemented function to rcmd build, check and install to a directory

### PR DESCRIPTION
**Pull request by maximilianmordig**
Closes issue #52

### CODE REVIEW

File: R/dependencies.R [hunk 1]
```diff
@@ -80,10 +80,10 @@ add_project_to_local_repos <- function(project, local_repos) {
 #' }
 #'
 install_deps <- function(project = ".", feature = NULL,
-                                  local_repos = get_local_pkgs_from_config(),
-                                  direction = "upstream",
-                                  install_project = TRUE, dry_install = FALSE, verbose = 0,
-                                  install_external_deps = TRUE, ...) {
+                         local_repos = get_local_pkgs_from_config(),
+                         direction = "upstream",
```
Comment (nikolas-burkoff) [inline R/dependencies.R:84]: should the direction argument be removed? With the change to `get_true_deps_graph(internal_deps, direction = "upstream")` below `direction` doesn't seem to be used is this right?

File: R/dependencies.R [hunk 2]
```diff
@@ -825,3 +825,85 @@ update_with_direct_deps <- function(project = ".",
...
+#' )
+#' build_check_install_repos(
+#'   list(list(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   host = "https://code.roche.com")),
+#'   feature = "master",
+#'   direction = "upstream",
+#'   #local_repos = data.frame(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   #host = "https://code.roche.com",
+#'   #directory = "../example_ecosystem/stageddeps.electricity/",
+#'   #stringsAsFactors = FALSE),
```
Comment (nikolas-burkoff) [inline R/dependencies.R:858]: don't think we should have commented out args in the example

File: R/dependencies.R [hunk 3]
```diff
@@ -825,3 +825,85 @@ update_with_direct_deps <- function(project = ".",
...
+#'   artifact_dir = "/tmp/test112"
+#' )
+#' }
+build_check_install_repos <- function(repos_to_process, feature,
```
Comment (nikolas-burkoff) [inline R/dependencies.R:862]: does this solve https://github.com/openpharma/staged.dependencies/issues/46? If so (or with a small tweak) then lets get that in as well?

File: R/dependencies.R [hunk 4]
```diff
@@ -825,3 +825,85 @@ update_with_direct_deps <- function(project = ".",
...
+
+  internal_deps <- rec_checkout_internal_deps(
+    repos_to_process, feature, direction, local_repos, verbose
+  )
+  deps <- get_true_deps_graph(internal_deps, direction = "upstream")
+  install_order <- get_install_order(deps[["upstream_deps"]])
+  install_order_paths <- vapply(
+    install_order, function(x) internal_deps[[hash_repo_and_host(x)]], character(1)
+  )
+
```
Comment (nikolas-burkoff) [inline R/dependencies.R:879]: This only builds, installs, tests upstream? I think we want this to do upstream and downstream dependent on direction? Looking at the issue:

"We would like to create a CI pipeline that installs and tests all packages in the staged dependencies and run a matrix test over different environments."

**Review by nikolas-burkoff — CHANGES_REQUESTED** [review]
A few comments to address on this one. Give me a shout if anything isn't clear

File: R/dependencies.R (hunk 1 above)
Comment (maximilianmordig) [reply R/dependencies.R:84]: `direction` is used on line 104. We need the true dependency graph, where arrows point upstream, so we don't pass `direction` here.

File: R/dependencies.R (hunk 4 above)
Comment (maximilianmordig) [reply R/dependencies.R:879]: No, but it uses the graph (where arrows point upstream) to compute the installation order. We can discover packages going upstream and/or downstream. However, to install, we only need the upstream graph.

File: R/dependencies.R (hunk 1 above)
Comment (maximilianmordig) [reply R/dependencies.R:84]: done, thanks

File: R/dependencies.R (hunk 4 above)
Comment (maximilianmordig) [reply R/dependencies.R:879]: I updated the doc.

File: R/dependencies.R (hunk 2 above)
Comment (maximilianmordig) [reply R/dependencies.R:858]: Did you mean to uncomment the `local_repos` arg? I did so now.

Comment (maximilianmordig) [reply R/dependencies.R:858]: Did you mean to uncomment the `local_repos` arg? I did so now.

File: R/dependencies.R (hunk 3 above)
Comment (maximilianmordig) [reply R/dependencies.R:862]: Added support for this via the `steps` argument.

File: R/dependencies.R [hunk 5]
```diff
@@ -852,6 +858,114 @@ update_with_direct_deps <- function(project = ".",
...
+  )
+  # we need the upstream direction (rather than variable direction) to
+  # compute the installation order
+  deps <- get_true_deps_graph(internal_deps, direction = "upstream")
+  install_order <- get_install_order(deps[["upstream_deps"]])
+  install_order_paths <- vapply(
+    install_order, function(x) internal_deps[[hash_repo_and_host(x)]], character(1)
+  )
+
+  dir.create(file.path(artifact_dir, "build_logs"))
```
Comment (nikolas-burkoff) [inline R/dependencies.R:929]: do we want to create these only if build/check where selected in step

File: R/dependencies.R (hunk 3 above)
Comment (nikolas-burkoff) [reply R/dependencies.R:906]: Should we default `feature` to `main`?

File: R/dependencies.R [hunk 6]
```diff
@@ -852,6 +858,114 @@ update_with_direct_deps <- function(project = ".",
...
+#' # also adds commit SHA
+#' install_deps(someArgs, direction = c("upstream", "downstream"))
+#' }
+build_check_install_repos <- function(repos_to_process, feature,
+                                      direction = c("upstream", "downstream"),
+                                      local_repos = get_local_pkgs_from_config(),
+                                      verbose = 0,
+                                      steps = c("build", "check", "install"),
+                                      rcmd_args = list(check = c("--no-manual")),
+                                      artifact_dir = tempfile()) {
```
Comment (nikolas-burkoff) [inline R/dependencies.R:912]: Do we want the default to be a tempfile()? When you run install or build within R Studio it puts the tar file in the parent direction of the package and at the sample place a folder  called <<package_name>>.00check (or something like that) to store the output of the check

File: R/dependencies.R [hunk 7]
```diff
@@ -852,6 +858,114 @@ update_with_direct_deps <- function(project = ".",
...
+#'   stringsAsFactors = FALSE)
+#' )
+#' build_check_install_repos(
+#'   list(list(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   host = "https://code.roche.com")),
+#'   feature = "master",
+#'   direction = "upstream",
+#'   local_repos = data.frame(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   host = "https://code.roche.com",
+#'   directory = "../example_ecosystem/stageddeps.electricity/",
```
Comment (nikolas-burkoff) [inline R/dependencies.R:895]: Not to change right now but this feels a little clunky as a way of giving the information something like this seems easier for users (not saying it should be this - just something to think about)

```
build_check_install_repos(c( "x/y@https://github.com"="remote", 
                                             "x/y2@https://github.com"="../y@local")
                                         )
```

Comment (maximilianmordig) [reply R/dependencies.R:895]: later (on cleanup)

**Review by nikolas-burkoff — APPROVED** [review]
Lets get this in.
As discussed for future 1) input arguments simpler for users 2) refactor to streamline code base

**Pull request merged**